llvm/test/CodeGen/AMDGPU/amdgpu-oclc-reflect.ll

```diff
@@ -0,0 +1,17 @@
+; RUN: opt -S -p amdgpu-oclc-reflect %s | FileCheck %s -check-prefixes=CHECK,CHECK-SAFE-ATOMICS
+; RUN: opt -S -p amdgpu-oclc-reflect -amdgpu-oclc-unsafe-int-atomics=true %s | FileCheck %s -check-prefixes=CHECK,CHECK-UNSAFE-ATOMICS
+
+target triple = "amdgcn-amd-amdhsa"
+
+@.str = private unnamed_addr addrspace(4) constant [31 x i8] c"AMDGPU_OCLC_UNSAFE_INT_ATOMICS\00", align 1
+
+declare hidden i32 @__oclc_amdgpu_reflect(ptr addrspace(4) noundef) local_unnamed_addr
+
+define i32 @foo() {
+; CHECK-NOT: call i32 @__oclc_amdgpu_reflect(ptr addrspace(4) noundef @.str)
+; CHECK-SAFE-ATOMICS: ret i32 0
+; CHECK-UNSAFE-ATOMICS: ret i32 1
+  %call = tail call i32 @__oclc_amdgpu_reflect(ptr addrspace(4) noundef @.str)
```

Comment:
I don't know how NVVMReflect works but I don't see what's stopping LLVM from optimizing two (or more) of these into a common call with a non-constant argument:

``` llvm
br i1 cond %c, label %if, label %else
if:
  %i0 = call @__oclc_amdgpu_reflect("hello")
  br label %merge
else:
 %i1 = call @__oclc_amdgpu_reflect("world")
  br label %merge
merge:
  %i = phi i32 [ i32 %i0, label %if ], [ i32 %i1, label %else ]
```

optimized to:

``` llvm
%arg = select i1 %c, ptr "hello", ptr "world"
%i = call @__oclc_amdgpu_reflect(%arg)
```

I don't know if there's a way around this with regular user functions, other than acknowledging it may happen. If it were an intrinsic we could probably fix this with function attributes (`IntrNoMerge` etc)

I don't know what the best way forward is. Making the pass less shouty (not so many `assert false`s but emitting warnings when encountering an unknown "reflection") might be one idea. We could assume and document that `0` is the default value, and make any issues into a performance/developer problem rather than a compiler crashing user-can't-compile-their-code problem.

Reply:
This is an interesting thought. I think since the pass at the moment should only be used by us in libclc, I would prefer a hard failure for an unknown arg or a non constant arg.

In theory the compiler could optimize some phi into a single call with a non constant arg, however the way NVVMReflect is used (since this pass is essentially just NVVMReflect for AMDGPU) is usually to control granular and unrelated compile time configurations like: "What is the arch?", "Do we want prec sqrt?", "Do we want this rounding mode?". As such it is unlikely that we have something like:
```llvm
if:
  %i0 = call @__oclc_amdgpu_reflect("FTZ_ON")
  br label %merge
else:
 %i1 = call @__oclc_amdgpu_reflect("PREC_SQRT")
  br label %merge
merge:
  %i = phi i32 [ i32 %i0, label %if ], [ i32 %i1, label %else ]
```
Moreover as implementers and library developers, the likely sole users of this pass, we are in the position to debug this pass if it does indeed fail, which should be clear from the `isa` assert within `casts`

Reply:
Yeah, I suspected as much - it's unlikely to be a problem in how we usage. My more principled side just wanted to chime in for a second, in that this isn't a robust general-purpose solution and maybe that's fine. I also get that we (may) want to "reflect" how the NVVMReflectPass works, or doesn't, in the same situations.